kern/gnutls_masterkey.h

```diff
@@ -0,0 +1,356 @@
+// struct gnutls_session_int {
+//     security_parameters_st security_parameters;
+//     record_parameters_st *record_parameters[MAX_EPOCH_INDEX];
+//     internals_st internals;
+//     gnutls_key_st key;
+// };
+// 
+// gnutls_session_int --> security_parameters_st
+// struct security_parameters_st {
+//     // ignore
+//     // ...
+//     const mac_entry_st *prf;
+//     uint8_t master_secret[GNUTLS_MASTER_SIZE];
+//     uint8_t client_random[GNUTLS_RANDOM_SIZE];
+//     // ignore
+//     // ...
+//     const version_entry_st *pversion;
+// };
+// 
+// gnutls_session_int --> security_parameters_st --> mac_entry_st
+// typedef struct mac_entry_st {
+//     // ignore
+//     // ...
+//     gnutls_mac_algorithm_t id;
+//     // ignore
+//     // ...
+// } mac_entry_st;
+// 
+// gnutls_session_int --> security_parameters_st -> version_entry_st
+// typedef struct {
+//     // ignore
+//     // ...
+//     gnutls_protocol_t id;	/* gnutls internal version number */
+//     // ignore
+//     // ...
+// } version_entry_st;
+// 
+// gnutls_session_int --> gnutls_key_st
+// struct gnutls_key_st {
+//     // ignore
+//     // ...
+//     union {
+//         struct {
+//             // ignore
+//             // ...
+//             uint8_t hs_ckey[MAX_HASH_SIZE]; /* client_hs_traffic_secret */
+//             uint8_t hs_skey[MAX_HASH_SIZE]; /* server_hs_traffic_secret */
+//             uint8_t ap_ckey[MAX_HASH_SIZE]; /* client_ap_traffic_secret */
+//             uint8_t ap_skey[MAX_HASH_SIZE]; /* server_ap_traffic_secret */
+//             uint8_t ap_expkey[MAX_HASH_SIZE]; /* {early_,}exporter_master_secret */
+//             // ignore
+//             // ...
+//         } tls13; /* tls1.3 */
+// 
+//         /* Follow the SSL3.0 and TLS1.2 key exchanges */
+//         struct {
+//             // ignore
+//             // ...
+//         } tls12; /* from ssl3.0 to tls12 */
+//     } proto;
+//     // ignore
+//     // ...
+// };
+//
+
+#define GNUTLS_RANDOM_SIZE 32
+#define GNUTLS_MASTER_SIZE 48
+#define MAX_HASH_SIZE 64
+
+struct gnutls_mastersecret_st {
+    u32 version;
+    /* from ssl3.0 to tls1.2 */
+    u8 client_random[GNUTLS_RANDOM_SIZE];
+    u8 master_secret[GNUTLS_MASTER_SIZE];
+
+    /* tls1.3 */
+    u32 cipher_id;
+    u8 client_handshake_secret[MAX_HASH_SIZE];
+    u8 server_handshake_secret[MAX_HASH_SIZE];
+    u8 client_traffic_secret[MAX_HASH_SIZE];
+    u8 server_traffic_secret[MAX_HASH_SIZE];
+    u8 exporter_master_secret[MAX_HASH_SIZE];
+};
+
+// #define GNUTLS_MAC_SHA256 6
+// #define GNUTLS_MAC_SHA384 7
+
+#define GNUTLS_SSL3 1
+#define GNUTLS_TLS1_0 2
+#define GNUTLS_TLS1 GNUTLS_TLS1_0
+#define GNUTLS_TLS1_1 3
+#define GNUTLS_TLS1_2 4
+#define GNUTLS_TLS1_3 5
+#define GNUTLS_DTLS1_0 201
+#define GNUTLS_DTLS1_2 202
+
+// Supported version: gnutls 3.7.9/3.8.1/3.8.3
+// Release: Debian 12/ubuntu 23.10/ubuntu 24.04
+
+// gnutls_session_int->security_parameters
+#define GNUTLS_SESSION_INT_SECURITY_PARAMETERS 0x0
```

Comment:
Do all versions of GNUTLS have the same offset address?

Do we need to determine the version of the class library and select the corresponding bytecode file, just like openssl?

Reply:
I not plan to support too much versions of GNUTLS, just the default version which installed in popular Linux distribution, this three version use same offset.